doc: rework section on subtyping (#5185)
dfinity/motoko: doc/md/fundamentals/2-types/14-subtyping.md

`None` is similar to an empty set, as it has no elements and is a subset of every other set `{} ⊆ A`.

## Any: the greatest type

In Motoko, `Any` contains all possible values. By definition, every type is a subtype of `Any`: for any type `T`, there is `T <: Any`.

This makes `Any` the greatest type in the subtype hierarchy.

A function that takes an argument of type `Any` can be applied to a value of any type, since all types are compatible with `Any`.

``` motoko no-repl
func discard(a : Any) {};

discard(0); // Allowed, since `Nat <: Any`
discard(true); // Allowed, since `Bool <: Any`
discard("abc"); // Allowed, since `Text <: Any`
```

`Any` is similar to the universal set `U` that contains all possible elements. Every set `A` is a subset of the universal set `A ⊆ U`.

## Options

If `T <: U`, then `?T <: ?U` because option subtyping is covariant. This means an [optional value](https://internetcomputer.org/docs/motoko/fundamentals/types/options-results) of a subtype can be used as an optional value of a supertype.

```motoko no-repl
let a : ?Nat = ?5;
let b : ?Int = a;     // Allowed, since `Nat <: Int` implies `?Nat <: ?Int`
```

> In Motoko, the literal `null` has the type `Null`, which is a subtype of any optional type. For any type `T`, there is `Null <: ?T`.

Author: ```suggestion
In Motoko, the literal `null` has the type `Null`, which is a subtype of any optional type. For any type `T`, `Null <: ?T`.
```

let nats : [Nat] = [1, 2, 3];
let ints : [Int] = nats;  // Allowed, since `Nat <: Int` we also have `[Nat] <: [Int]`.
```

## Mutable arrays

[Mutable arrays](https://internetcomputer.org/docs/motoko/fundamentals/types/mutable-arrays) of the form `[var T]` do not support interesting subtyping.
The mutable array constructor `[var T]` is invariant in `T`.

This means that `[var T]` is a subtype of `[var U]` only if `T` and `U` are equivalent types, that is, both `T <: U` and `U <: T` must hold.

Allowing `[var T] <: [var U]` whenever `T <: U` (without requiring `U <: T`) would be unsafe, as it could allow writing values of type `U` into an array that only expects values of type `T`.

```motoko no-repl
let nats : [var Nat] = [var 1, 2, 3];
let ints : [var Int] = nats;  // Not allowed, because `[var Nat] </: [var Int]`.
```

## Functions

[Functions](https://internetcomputer.org/docs/motoko/fundamentals/types/functions) also support subtyping.
A function type `T1 -> T2` is a subtype of another function type `U1 -> U2` provided that:

1. The argument types are related in the opposite direction (`U1 <: T1` ).
2. The return types are related in the same direction (`T2 <: U2`).

> This means that function subtyping is contravariant in the arguments of the function and covariant in the results.
> 
> This might seems confusing at first, but the intuition is quite simple.
> A function can be used at a supertype provided that type requires a more specific argument and produces a more general result.
> This ensures that the orginal function, when used at the supertype,
> will only receive arguments it can consume (since `U1 <: T1`) and produce results than are expected (since `T2 <: U2`).

Author: ```suggestion
Function subtyping is contravariant in the argument types and covariant in the return type.

A function of type `T1 → T2` can be used where a function of type `U1 → U2` is expected only if `U1 <: T1` (the expected argument is more specific) and `T2 <: U2` (the actual result is more generic).

This ensures safety because the function will never be passed arguments it can’t handle, and it will always produce results that meet the expectations of the surrounding context.
```

```motoko
let n : Nat = 5;
let i : Int = n;        // Allowed, since `Int <: Nat`
```

```motoko
let i : Int = -5;
let n : Nat = i;        // Not allowed, since `Int </: Nat`
```

## None: the least type

`None` is an empty type in Motoko. It contains no values at all so, and, by definition, `None` is a subtype of every other type: `None <: T`, for any `T`.

None is the least (as in smallest) type.

You might think the `None` type is useless, but it is used to type expressions that never produce a value, like the infinite loop `loop {}`.

``` motoko no-repl
func impossible() : None { loop {} };

impossible() # impossible();  // Allowed, since `None <: Text`
impossible() + impossible();  // Allowed, since `None <: Nat`

if (false == true) impossible(); // Allowed, since `None <: ()`
```

> Call the function `impossible()` will force your program to enter an infinite loop, returning no value at all.
> Because `None` is the least type, it can be used wherever any other type is expected.
> 
> `None` is similar to the empty set: the empty set has no elements and is a subset of every other set `{} ⊆ A`.

Reviewer: ```suggestion
Calling the function `impossible()` will force the program to enter an infinite loop and return no value.
Because `None` is the least type, it can be used wherever any other type is expected.

`None` is similar to an empty set, as it has no elements and is a subset of every other set `{} ⊆ A`.
```

## Any: the greatest type

> Motoko's `Any` type contains all values. By definition, every type is a subset of `Any`, `T <: Any`.
> `Any` is the greatest (as in largest) type in Motoko.
> 
> A function that accepts an `Any` argument can be applied to "any" type of argument.

Reviewer: ```suggestion
In Motoko, `Any` contains all possible values. By definition, every type is a subtype of `Any`: for any type `T`, there is `T <: Any`.

This makes `Any` the greatest type in the subtype hierarchy.

A function that takes an argument of type `Any` can be applied to a value of any type, since all types are compatible with `Any`.
```

``` motoko no-repl
func discard(a : Any) {};

discard(0); // Allowed, since `Nat <: Any`
discard(true); // Allowed, since `Bool <: Any`
discard("abc"); // Allowed, since `Text <: Any`
```

> `Any` is similar to the universal set `U`, the set that contains all possible elements.
> Every set `A` is a subset of the universal set, `A ⊆ U`.

Reviewer: ```suggestion
`Any` is similar to the universal set `U` that contains all possible elements. Every set `A` is a subset of the universal set `A ⊆ U`.
```

## Options

> If `T <: U`, then `?T <: ?U` (option subtyping is covariant). This means an [optional value](https://internetcomputer.org/docs/motoko/fundamentals/types/options-results) of a subtype can be used as an optional value of a supertype.

Reviewer: ```suggestion
If `T <: U`, then `?T <: ?U` because option subtyping is covariant. This means an [optional value](https://internetcomputer.org/docs/motoko/fundamentals/types/options-results) of a subtype can be used as an optional value of a supertype.
```

```motoko no-repl
let a : ?Nat = ?5;
let b : ?Int = a;     // Allowed, since `Nat <: Int` implies `?Nat <: ?Int`
```

> In Motoko, the literal `null` has type `Null` which subtypes any option type, that is `Null <: ?T` (for any `T`).
> This means you can use `null` as the absent value of any optional type.

Reviewer: ```suggestion
In Motoko, the literal `null` has the type `Null`, which is a subtype of any optional type. For any type `T`, there is `Null <: ?T`.

This means `null` can be used as the absent value for any optional type.
```

```motoko no-repl
let n : Null = null;
let oi : ?Int = n;      // Allowed, since `Null <: ?Int`
let ot : ?Text = n;     // Allowed, since `Null <: ?Text`
```

## Records and objects

> [Records](https://internetcomputer.org/docs/motoko/fundamentals/types/records) and, more generally,  [objects](https://internetcomputer.org/docs/motoko/fundamentals/types/objects-classes)

Reviewer: ```suggestion
[Records](https://internetcomputer.org/docs/motoko/fundamentals/types/records) and, more generally, [objects](https://internetcomputer.org/docs/motoko/fundamentals/types/objects-classes)
```

support subtyping, both in the required fields and the types of those fields.

An object type `T` is a subtype of another object type `U`, if `T` requires all the fields required by `U`,
with field types that are subtypes of those in `U`.
Note that `T` may provide more fields than `U`.

```motoko no-repl
type A = { name : Text; age : Nat };
type B = { name : Text; age : Int };
type C = { name : Text };

let a : A = { name = "Ghost"; age = 25 };
let b : B = { name = "Other"; age = -1 };
let c : C = { name = "Motoko" };

let b1 : B = a; // Allowed since `age : Nat` implies `age : Int` (`Nat <: Int`).
let c1 : C = b; // Allowed, since `B` has all fields of `C`
let a1 : A = b; // Not allowed, since `age : Int` does not imply `age : Nat` (`Int </: Nat`).
let b2 : B = c; // Not allowed, because `C` is missing field `age`.
```

> If the field of an object is mutable (e.g. `var age : Nat`), then any field in the supertype must also be mutable, with equivalent content.

Reviewer: ```suggestion
If the field of an object is mutable (e.g., `var age : Nat`), then any field in the supertype must also be mutable with equivalent content.
```

> For example, consider these object types with mutable `age` fields:

Reviewer: ```suggestion
For example, consider the following object types with mutable `age` fields:
```

```motoko no-repl
type A = { name : Text; var age : Nat };
type B = { name : Text; var age : Int };
```

> Now neither `A <: B` nor `B <: A` since the contents of the mutable `age` field, `Nat` and `Int`, are not equivalent.
> However, we still have `A <: C` and `B <: C` (dropping the `age` field entirely).

Reviewer: ```suggestion
Now, neither `A <: B` nor `B <: A` holds, because the `age` fields differ in type because `Nat` and `Int` are not equivalent.

However, both `A` and `B` are still subtypes of `C`, since `C` lacks the `age` field entirely and subtyping allows dropping fields.
```

```motoko no-repl _include=Days
let d : Day = #mon;
let wd : WeekDay = d;  // Not allowed, since `Day </: WeekDay` (`d` could also be `#sat`)
```

Variants with arguments can also be related by subtyping:

```motoko no-repl
type Ok<T> = {#ok : T};
type Err<E> = {#err : E};
type Result<T, E> = {#ok : T; #err : E};

let err : Err<Text> = #err "ohoh";
let ok : Ok<Nat> = #ok 0;
let rerr : Result<Int, Text> = err;  // Allowed, since `{#err : Text} <: {#ok : Int; #err : Text}`,
                                     // that is, `Err<Text> <: Result<Int, Text>`
let rok : Result<Int, Text> = ok; // Allowed, since `{#ok : Nat} <: {#ok : Int; #err : Text}`,
                                  // that is, `Ok<Nat> <: Result<Int, Text>`
```

## Immutable arrays

[Immutable arrays (`[T]`)](https://internetcomputer.org/docs/motoko/fundamentals/types/immutable-arrays) support covariant subtyping, meaning if `T <: U`, then `[T] <: [U]`.

If `T <: U`, then an (immutable) array of type `[T]` can be used as an array of type `[U]`.

```motoko no-repl
let nats : [Nat] = [1, 2, 3];
let ints : [Int] = nats;  // Allowed, since `Nat <: Int` we also have `[Nat] <: [Int]`.
```

## Mutable arrays

> [Mutable arrays]](https://internetcomputer.org/docs/motoko/fundamentals/types/mutable-arrays) of the form  `[var T]` do not support interesting subtyping.

Reviewer: ```suggestion
[Mutable arrays](https://internetcomputer.org/docs/motoko/fundamentals/types/mutable-arrays) of the form `[var T]` do not support interesting subtyping.
```

> The mutable array constructor `[var T]` is invariant in `T`.
> This means that `[var T] <: [var U]` only when `T` and `U` are equivalent.
> 
> Allowing `[var T] <: [var U]` whenever `T <: U`, without requiring `U <: T` would not be safe.

Reviewer: ```suggestion
The mutable array constructor `[var T]` is invariant in `T`.

This means that `[var T]` is a subtype of `[var U]` only if*`T` and `U` are equivalent types, that is, both `T <: U` and `U <: T` must hold.

Allowing `[var T] <: [var U]` whenever `T <: U` (without requiring `U <: T`) would be unsafe, as it could allow writing values of type `U` into an array that only expects values of type `T`.
```

```motoko no-repl
let nats : [var Nat] = [var 1, 2, 3];
let ints : [var Int] = nats;  // Not allowed, because `[var Nat] </: [var Int]`.
```

> If subtyping between mutable arrays were allowed, then the legal assignment
> 
> ``` motoko no-repl
> ints[0] := -1;
> ```
> 
> would have the side-effect of also setting `nats[0]` to `-1`, but `nats` must only contain `Nat`s!

Reviewer: ```suggestion

```

Reviewer: Since this is not allowed, this is not necessary to call out.

Author: (Reapplied manually - somehow your suggestion was lost).

## Functions

> [Function](https://internetcomputer.org/docs/motoko/fundamentals/types/functions) also support subtyping.

Reviewer: ```suggestion
[Functions](https://internetcomputer.org/docs/motoko/fundamentals/types/functions) also support subtyping.
```

> A function type `T1 -> T2` is a subtype of another function type `U1 -> U2` provided

Reviewer: ```suggestion
A function type `T1 -> T2` is a subtype of another function type `U1 -> U2` provided that:
```

> 1. `U1 <: T1` : the argument types are related in the opposite direction.
> 2. `T2 <: U2` :  the return types are related in the same direction.

Reviewer: ```suggestion
1. The argument types are related in the opposite direction (`U1 <: T1` ).
2. The return types are related in the same direction (`T2 <: U2`).
```

> This means that function subtyping is contravariant in the arguments of the function and covariant in the results.
> 
> This might seems confusing at first, but the intuition is quite simple.
> A function can be used at a supertype provided that type requires a more specific argument and produces a more general result.
> This ensures that the orginal function, when used at the supertype,
> will only receive arguments it can consume (since `U1 <: T1`) and produce results than are expected (since `T2 <: U2`).
> 
> As a simple example, consider the `magnitude` function that returns the absolute value, or magnitude, of an integer.

Reviewer: ```suggestion
Function subtyping is contravariant in the argument types and covariant in the return type.

A function of type `T1 → T2` can be used where a function of type `U1 → U2` is expected only if `U1 <: T1` (the expected argument is more specific) and `T2 <: U2` (the actual result is more generic).

This ensures safety because the function will never be passed arguments it can’t handle, and it will always produce results that meet the expectations of the surrounding context.

As a simple example, consider the `magnitude` function that returns the absolute value, or magnitude, of an integer.
```

Author: (this suggestion was lost, I've reapplied manually)

```motoko no-repl name=magnitude
import Int "mo:base/Int"
func magnitude(i : Int) : Nat { Int.abs(i) };
```

Its most precise type is `Int -> Nat`, but, due to subtyping, it also has types `Nat -> Nat` (making the argument more specific),
`Int -> Int`  (making the result more general) and `Int -> Int` (doing both).

``` motoko no-repl _include=magnitude
let i2n : Int -> Nat = magnitude;
let n2n : Nat -> Nat = magnitude;
let i2i : Int-> Int = magnitude;
let n2i : Nat-> Int = magnitude;
```

> You may be wondering what happens if the one or both of the function types is generic and has type parameters.
> Motoko uses the simple rule that two function types with type parameters are subtypes only when both types have the same parameters
> (up to renaming), and the arguments and results are related according to 1. and 2. above.

Reviewer: ```suggestion
You might wonder what happens when one or both of the function types are generic and have type parameters.

Motoko applies a straightforward rule: two generic function types are considered subtypes only if they have the same type parameters (up to renaming), and their parameter and result types follow the usual subtyping rules.
```

## Modules and actors

> Modules and actors support subtyping just like objects, meaning a module(actor) is a subtype of a similar module(actor) with some fields removed,
> provided the types of the fields in common are related by subtyping.
> 
> For actors, which can only contain shared functions as fields, this means that you can remove some functions or replace the types of common
> functions with supertypes.

Reviewer: ```suggestion
Modules and actors in Motoko support subtyping in the same way as objects.

Specifically, a module (or actor) is a subtype of another module (or actor) with some fields removed, as long as the remaining shared fields have types that are related by subtyping.

In the case of actors, which can only contain shared functions as fields, this means:

* You can remove some functions.
* You can replace the types of shared functions with supertypes (according to function subtyping rules).
```

## Recursive and generic types

[Recursive and generic types](https://internetcomputer.org/docs/motoko/fundamentals/types/advanced-types) can be subtypes of each other when their definitions allow it.

> Consider these two recursive point types, the second one adding a `color` field:

Reviewer: ```suggestion
Consider the following recursive point types, where the second extends the first by adding a color field:
```

```motoko no-repl name=Point
type Point = {
  x : Int;
  move: Int -> Point
};

type ColorPoint = {
  color : {#red; #blue};
  x : Int;
  move: Int -> ColorPoint
};
```

`ColorPoint` is a subtype of `Point`:

``` motoko no-repl _include=Point
let cp : ColorPoint = {
  color = #red;
  x = 0;
  move = func (dx : Int) { { cp with x = cp.x + dx } }
};
let c : Point = cp;
```

> This also works for recursive variants and even generic types:

Reviewer: ```suggestion
This also works for recursive variants and even generic types.
```
